### PR TITLE
 Fix handling of NaN for map variants

### PR DESCRIPTION
**Pull request by bikramSingh91**
Summary:
Highlights of the this change:
- Introduces a utility functors for floating point types that provide
comparator and hash functor to implement consistent behavior or NaNs
across the codebase. These can be passed to standard containers and
functions like std::map and std::sort, etc.
- These utility functors will be used in upcoming changes to fixes for NaN
behavior across the codebase wherever folly or std containers or
functions are used that can accept such functors.
- Fix NaN handling of floating points for map variants which uses a
std::map to hold key and value variants. Without this, using NaN as
a key would result in an inconsistent state because NaN cannot be
compared with other values. This resulted in inconsistent comparison
of map vectors in tests where QueryAssertions.h uses a map variant
to compare such vectors. (See asserEqualResults())

Differential Revision: D57187815

### CODE REVIEW

File: velox/type/FloatingPointUtil.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+/// Custom comparator and hash functors for floating point types. These are
+/// designed to ensure consistent NaN handling according to the following rules:
+/// - NaN == NaN returns true, even for NaNs with differing binary
+///   representations.
+/// - NaN is considered greater than infinity.
+/// These can be passed to standard containers and functions like std::map and
+/// std::sort, etc.
+namespace util::floating_point {
+template <typename FLOAT>
```
Comment (kgpai) [inline velox/type/FloatingPointUtil.h:33]: Maybe add this to ensure this is only valid for floating point types ?

```

template <typename FLOAT, typename std::enable_if<std::is_floating_point<T>::value >::type* = 0>  
..

```

File: velox/type/FloatingPointUtil.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+#include <array>
+#include <cmath>
+#include <vector>
+
+namespace facebook::velox {
+
+/// Custom comparator and hash functors for floating point types. These are
+/// designed to ensure consistent NaN handling according to the following rules:
+/// - NaN == NaN returns true, even for NaNs with differing binary
```
Comment (kgpai) [inline velox/type/FloatingPointUtil.h:27]: Another question, if -NaN == NaN , then is sorting on these values deterministic or indeterministic (if you dont knwo the order of the Nan's).

File: velox/type/Variant.cpp [hunk 3]
```diff
@@ -670,7 +671,59 @@ variant variant::create(const folly::dynamic& variantobj) {
...
+template <TypeKind KIND>
+bool variant::lessThan(const variant& a, const variant& b) const {
+  using namespace facebook::velox::util::floating_point;
+  if (a.isNull() && !b.isNull()) {
+    return true;
+  }
+  if (a.isNull() || b.isNull()) {
+    return false;
+  }
+  if constexpr (KIND == TypeKind::REAL) {
```
Comment (kgpai) [inline velox/type/Variant.cpp:683]: nit : you can get this into one if , with : 

``` 

 if constexpr (td::is_floating_point_v<KIND>)  {
  return NaNAwareLessThank<KIND>{}...
  
  }
  
  ```

File: velox/type/Variant.cpp [hunk 4]
```diff
@@ -670,7 +671,59 @@ variant variant::create(const folly::dynamic& variantobj) {
...
+    return NaNAwareEquals<float>{}(a.value<KIND>(), b.value<KIND>());
+  }
+  if constexpr (KIND == TypeKind::DOUBLE) {
+    return NaNAwareEquals<double>{}(a.value<KIND>(), b.value<KIND>());
+  }
+  // todo(youknowjack): centralize equality semantics
```
Comment (kgpai) [inline velox/type/Variant.cpp:711]: maybe remove this todo ?

File: velox/type/tests/VariantTest.cpp [hunk 5]
```diff
@@ -197,6 +197,33 @@ TEST(VariantTest, equalsWithEpsilonFloat) {
   ASSERT_FALSE(variant(sum1).equalsWithEpsilon(variant(sum3)));
 }
 
+TEST(VariantTest, mapWithNaNKey) {
```
Comment (kgpai) [inline velox/type/tests/VariantTest.cpp:200]: Can you also add a test for multimap_agg to validate ? Thanks !

**Review by kgpai — COMMENTED** [review]
Thanks, Looks great, some minor nits.

File: velox/type/FloatingPointUtil.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#pragma once
+
+#include <array>
+#include <cmath>
+#include <vector>
+
+namespace facebook::velox {
+
+/// Custom comparator and hash functors for floating point types. These are
+/// designed to ensure consistent NaN handling according to the following rules:
```
[the remark this answers is not in the document]
Comment (bikramSingh91) [reply velox/type/FloatingPointUtil.h:26]: This is for any kind of NaN (both signalling and quiet ranges), not just quiet NaNs

File: velox/type/FloatingPointUtil.h (hunk 2 above)
Comment (bikramSingh91) [reply velox/type/FloatingPointUtil.h:27]: In IEEE-754 there is no notion of a negative NaN, the sign bit exists and some programs print it out, but for all intents and purposes a NaN is just a Not-a-Number. There exists a classification of NaNs (signalling and quiet) but those specify range of values, but the sign is of no consequence.

To clarify, the `-` in this comment is a dash, not a negative sign. If that caused any confusion, I can use something like `->` instead

File: velox/type/FloatingPointUtil.h [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+/// - NaN == NaN returns true, even for NaNs with differing binary
+///   representations.
+/// - NaN is considered greater than infinity.
```
[the remark this answers is not in the document]
Comment (bikramSingh91) [reply velox/type/FloatingPointUtil.h:29]: Will add that as a `Note:`

File: velox/type/tests/VariantTest.cpp (hunk 5 above)
Comment (bikramSingh91) [reply velox/type/tests/VariantTest.cpp:200]: The fix for multimap_agg is in a follow up PR #9769

File: velox/type/FloatingPointUtil.h (hunk 2 above)
Comment (kgpai) [reply velox/type/FloatingPointUtil.h]: I was under impression it was a minus 😄

**Review by kgpai — APPROVED** [review]

**Pull request closed without merging**